### ISSUE TITLE
Select SHACL Compact Syntax editors

### ISSUE THREAD

**Issue by HolgerKnublauch** [w3c/data-shapes#187]
Moving towards a potential 1.2 release, the SHACL Compact Syntax needs new editors.

**Comment by jeswr** [comment]
I would be open to taking on an editing role for this spec.

Note that my current academic affiliation is not currently a paying W3C member - so I suspect I would only be able to do so as an invited expert (https://www.w3.org/invited-experts/).

**Comment by VladimirAlexiev** [comment]
I'd be happy to help @jeswr since I think that SHACLC is a crucial feature to compete with SHEX's dedicated syntax.
Some examples (with shots) in https://github.com/VladimirAlexiev/shaclc-mode.

**Comment by VladimirAlexiev** [comment]
@jeswr , I dug out ["shaclc" issues in the TopQuadrant repo](https://github.com/TopQuadrant/shacl/issues?q=is%3Aissue+shaclc):

- https://github.com/TopQuadrant/shacl/issues/98 (escape into turtle, and some extra links and ideas)
- https://github.com/TopQuadrant/shacl/issues/142 (test case)
- https://github.com/TopQuadrant/shacl/issues/91 (test case)
- https://github.com/TopQuadrant/shacl/issues/92 (test case)

Implementations:
- https://jena.apache.org/documentation/shacl/#shacl-compact-syntax
- (TQ SHACLC was retired after the Jena implementation appeared)
- https://github.com/jeswr/shaclcjs of @jeswr
- https://www.allotrope.org/ is a big foundation in the life sciences / scientific data.
  - https://gitlab.com/allotrope-open-source/shape-editor by @tw-osthus (Thomas Weber of OSTHUS). supports editing of shacl and shaclc files. Last updated 2019
  - https://gitlab.com/allotrope-open-source/shape-editor/-/blob/master/src/com.osthus.shapes.shaclc.parent/com.osthus.shapes.shaclc/src/com/osthus/shapes/shaclc/SHACLC.xtext shows they extend SHACLC with rules, incl SPARQL...

Related issues here:
- https://github.com/w3c/shacl/issues/7
- https://github.com/w3c/data-shapes/issues/179
- https://github.com/w3c/data-shapes/issues/186

**Comment by VladimirAlexiev** [comment]
- @jeswr Do you want to start a "scope" document with the info from the previous comment?
And of course, add your considerations to it. 
- @tw-osthus do you want to participate? You're the author of 1 of only 3 SHACLC implementations known to me

**Comment by jeswr** [comment]
> Do you want to start a "scope" document with the info from the previous comment?

Is this something that should be [Respec](https://github.com/w3c/respec)'d (in which case I'm guess it should be on this repo and I would need to be given the permissions to do that); or something that is sufficient to be thrown together in a HackMD/google doc/etc. for now?

**Comment by VladimirAlexiev** [comment]
I think it should be on github: for now as a markdown doc.

**Comment by HolgerKnublauch** [comment]
Please open a PR when ready. Once the process has started we can set up write permissions for the editors. (We do have two volunteers for the role of co-chairs, hopefully more on that this week).

**Comment by mielvds** [comment]
I'm interested in participating, but not sure in what way. I love the idea of SHACLC and it should be part of the spec (like @VladimirAlexiev points out: inherit the advantages of shex). I always wanted to use it, but I currently don't because of the unclear status and the minimal spec (e.g., it's unclear whether you can add comments, but you probably can?). 

I was not involved in the development of SHACL whatsoever, so I'm in the dark wrt this WG/CG, but I can review, give input, provide use cases, or be editor, whatever makes the most sense and leads to a reaslistic workload.

**Comment by tw-osthus** [comment]
I really liked SHACLC and would have preferred using it to write shapes.
So I made some advertising within the Allotrope Foundation to adopt it for the data models 5 years ago. Mainly this was the reason, I wrote the SHACLC editor at that time, but because it was "unofficial" it was not accepted by the community, which I can understand as a decision in a standardization body.

SHACL in turtle is not too bad, and I have got used to it, so I have dropped my work on SHACLC.

**Comment by VladimirAlexiev** [comment]
@tw-osthus Never give up! :-)
- Assembler is much better than Machine language.
But that's not enough reason to avoid high-level languages.
- Turtle is much better than eg RDF/XML.
But using it to write shapes or SPIN's SPARQL representation in triples, feels like writing in Assembler.

In any case, I'm sure we can ask you questions about your `xtext` grammar?
I'm not familiar with `xtext`, is there a tool to convert it to EBNF?

**Comment by afs** [comment]
JavaCC: https://github.com/apache/jena/blob/main/jena-shacl/shaclc/shaclc.jj (there are a few extensions and they are marked e.g. in `propertyParam`). The grammar is LL(1), that is, quite simple parser technology. 

There is elsewhere some scripts to translate the SPARQL javacc-checked grammar to EBNF and HTML and should be adaptable.

There is comprehensive tooling that Gregg Kellogg provides for W3C-favour BNF, including producing ReSpec-friendly HTML.

**Comment by VladimirAlexiev** [comment]
OK, so we got some grammar resources to reconcile:
- ANTLR: https://w3c.github.io/shacl/shacl-compact-syntax/#grammar-section by @HolgerKnublauch 
- JavaCC: https://github.com/apache/jena/blob/main/jena-shacl/shaclc/shaclc.jj by @afs 
- XTEXT: https://gitlab.com/allotrope-open-source/shape-editor/-/blob/master/src/com.osthus.shapes.shaclc.parent/com.osthus.shapes.shaclc/src/com/osthus/shapes/shaclc/SHACLC.xtext by @tw-osthus 
- Jison: https://github.com/jeswr/shaclcjs/blob/main/lib/shaclc.jison by @jeswr (is this the closest you have to a grammar?)

As for grammar visualization, I really like http://bottlecaps.de/rr/ui. 
I've applied it to SPARQL, SPARQL-Star, XSPARQL (you can't use it without such diagrams!) and Holger's SHACLC:
https://github.com/VladimirAlexiev/grammar-diagrams#shaclc
![image](https://github.com/w3c/shacl/assets/536250/508dbba1-2249-4d7d-a589-11d80e75bc06)

**Comment by jeswr** [comment]
> Jison: https://github.com/jeswr/shaclcjs/blob/main/lib/shaclc.jison by @jeswr (is this the closest you have to a grammar?)

Yes it is

> As for grammar visualization, I really like http://bottlecaps.de/rr/ui.

That link is broken for me

**Comment by VladimirAlexiev** [comment]
Thanks for notifying me!
- I checked 5 historic versions at archive.org, but they all return various rendering errors.
- I posted at https://github.com/GuntherRademacher/rr 
- But: "RR also runs on https://rr.red-dove.com/ui"
- https://github.com/GuntherRademacher/ebnf-convert is another project by the same guy that can help us unify the existing grammars to [W3C-style EBNF](http://www.w3.org/TR/2010/REC-xquery-20101214/#EBNFNotation)

@HolgerKnublauch should `shaclc` have its own repo, or should I make a folder in here?

**Comment by HolgerKnublauch** [comment]
A folder here should be fine but isn't the shacl-compact-syntax folder already a suitable place? It does contain a .g4 file.

**Comment by VladimirAlexiev** [comment]
@HolgerKnublauch thanks for the quick reply! See https://github.com/w3c/shacl/pull/49 
(also @jeswr: maybe you can let Jesse merge PRs to reduce your load?)

**Comment by HolgerKnublauch** [comment]
Yeah, I cannot look after the Compact Syntax anymore and would prefer if someone else can own this.

**Comment by jeswr** [comment]
Happy to take care of reviewing / approving PRs for SHACLC of given permission to do so.

@VladimirAlexiev did you mean to ping me instead of the user JesseW in your latest PR?

**Comment by VladimirAlexiev** [comment]
Yep: sorry: fixed.

Posted https://github.com/w3c/data-shapes/issues/193

**Comment by VladimirAlexiev** [comment]
hi @jeswr are you in the new WG? We missed you at the meeting today.
And are you still volunteering to be the main Compact editor?

**Comment by jeswr** [comment]
Yes, I'm in the WG - I had a meeting conflict today, which meant I couldn't turn up.

I have a large number of commitments at present, so right now I can commit no more than ~2 hours per week. If that is sufficient for primary editor for SHACL-C then I am happy to take on the role.

**Comment by VladimirAlexiev** [comment]
@nicholascar what do you say?

**Comment by HolgerKnublauch** [comment]
Will be handled as part of #217. For now I'd like to close this ticket although there has been some other discussion that doesn't seem to be about selecting the editor. Maybe those who remember can raise them as separate tickets?

**Comment by VladimirAlexiev** [comment]
Yeah, there's a gazillion comments here, but I think all the important ones (about various grammars and issues related to grammars) are captured in other issues and PRs.
